### ISSUE TITLE
How to set G Traing times and D Training times??

### ISSUE THREAD

**Issue by hongge831** [junyanz/pytorch-CycleGAN-and-pix2pix#357]
@andyli @strob @phillipi @taesung89 @iver56  Hi guys, I am just training your model with my own datasets,  I noticed that the model will train G 5 times then train the D one time?  how can I adjust how many times the training G and training D?? hope your reply thanks a lot!

**Comment by junyanz** [comment]
The current model trains G and D once.

**Comment by hongge831** [comment]
@junyanz thanks for quick reply. And can I set the times of training G and D manually? if yes ,how can I set the parameters on your codabase? **And another question**  Are your Cycle GAN and Pix2pix model based on by WGAN or the normal GAN, I am a freshman, would you mind telling me whether the WGAN will be better?

**Comment by hongge831** [comment]
@junyanz  on question one, I mean how can I set times of training G and D alternately(e.g. I want to train G2 2 times then train D 1 time). I am looking forward your reply. Thanks a lot.

**Comment by hongge831** [comment]
@junyanz thanks for reply. And I use your pix2pix model to train my own datasets, I use resnet-9blocks generator, but I get poor result. and my G-loss always large than D-loss.  there is my screenshot of training log. **would you mind telling me some tricks for deal with this situation?** thanks a lot :)

`(epoch: 14, iters: 58, time: 0.069, data: 0.001) G_GAN: 7.849 G_L1: 11.972 D_real: 0.006 D_fake: 0.001
(epoch: 14, iters: 158, time: 0.114, data: 0.001) G_GAN: 12.429 G_L1: 10.823 D_real: 0.001 D_fake: 0.000
(epoch: 14, iters: 258, time: 0.084, data: 0.001) G_GAN: 13.875 G_L1: 19.624 D_real: 0.000 D_fake: 0.000
(epoch: 14, iters: 358, time: 0.114, data: 0.001) G_GAN: 9.193 G_L1: 13.014 D_real: 0.000 D_fake: 0.000
(epoch: 14, iters: 458, time: 0.126, data: 0.001) G_GAN: 14.895 G_L1: 16.324 D_real: 0.000 D_fake: 0.000
(epoch: 14, iters: 558, time: 0.091, data: 0.001) G_GAN: 12.620 G_L1: 18.035 D_real: 0.000 D_fake: 0.000
`

**Comment by junyanz** [comment]
It seems that the D is too strong (or the G is too weak). D dominates the two-player game from the beginning. I will try to increase the capacity of G or decrease the capacity of D.

**Comment by hongge831** [comment]
@junyanz thanks for your reply. Would you mind telling me some specific tricks to deal with the problem? (e.g. how to increase the capacity of G or decrease the capacity of D?) I am freshman on GAN field. Thanks again for your kind reply and your nice project.

**Comment by junyanz** [comment]
just add/remove a few layers or increase/decrease the number of features per layer. You can try `ngf` and `ndf` first. See the [options](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/options/base_options.py#L20) for more details.

**Comment by hongge831** [comment]
thanks for your reply. I have another question should I set the `retain_graph=True` in the backward() when I want to train G more times than training D?  I found that when I just set as follows in `pix2pix_model.py`  (it can not work):
`    def optimize_parameters(self):


        self.forward()
        # update D
        self.set_requires_grad(self.netD, True)
        self.optimizer_D.zero_grad()
        self.backward_D()
        self.optimizer_D.step()
        for i in range(20):
            # update G
            self.set_requires_grad(self.netD, False)
            self.optimizer_G.zero_grad()
            self.backward_G()
            self.optimizer_G.step()`

**Comment by junyanz** [comment]
I think you can leave `self.set_requires_grad(self.netD, False)` outside the for loop.

**Comment by GXXTLRR** [comment]
cool
